### PR TITLE
Implement TrustedRandomAccess for slice::{Chunks, ChunksMut, Windows}

### PR DESCRIPTION
**Pull request by sdroege**
As suggested by @bluss in https://github.com/rust-lang/rust/issues/47115#issuecomment-354888334

### CODE REVIEW

**Comment by kennytm** [comment]
Thanks!

@bors r+

**Comment by sdroege** [comment]
Thanks for the fast review!

This is going to fail because of https://github.com/rust-lang/rust/pull/47113
I'll update it in a bit

**Comment by kennytm** [comment]
Okay. @bors r-

**Comment by sdroege** [comment]
For good measure, I'll also add unit tests for zipping the 3 iterators. It seems like this is not covered by any of the existing unit tests yet

**Comment by sdroege** [comment]
This will now fail until #47113 is merged for real

**Comment by sdroege** [comment]
Rebased on top of latest master

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -2228,6 +2236,16 @@ impl<'a, T> ExactSizeIterator for Chunks<'a, T> {}
 #[unstable(feature = "fused", issue = "35602")]
 impl<'a, T> FusedIterator for Chunks<'a, T> {}
 
+#[doc(hidden)]
+unsafe impl<'a, T> TrustedRandomAccess for Chunks<'a, T> {
+    unsafe fn get_unchecked(&mut self, i: usize) -> &'a [T] {
+        let start = i * self.chunk_size;
+        let end = cmp::min(start + self.chunk_size, self.v.len());
```
Comment (bluss) [inline src/libcore/slice/mod.rs]: There is a potential wraparound here isn't there? i using bounds but start + size can still wrap around. In particular for `[(); !0].chunks(!0/2 +1)` I think.

Comment (bluss) [reply src/libcore/slice/mod.rs]: I mean that "i is in bounds"

Comment (bluss) [reply src/libcore/slice/mod.rs]: It turns out Rust ICES on that array. For those that don't like my mobile-friendly syntax, `!0` is the same as `usize::max_value()`.

However Rust accepts a vec of this length, and we can demo [(playground link)](https://play.rust-lang.org/?gist=122589ec765d923f36890c23dab0f0d7&version=nightly)

```rust
fn main() {
    let v = vec![(); !0];
    let iter = v.chunks((!0)/2 +1);
    for elt in iter {
        println!("Found chunk of len {}", elt.len());
    }
}
```

Which has output:

```
Found chunk of len 9223372036854775808
Found chunk of len 9223372036854775807
```

**Comment by sdroege** [comment]
Good point, thanks. This should probably use `checked_add` and if it overflows then just use the length.

Will fix in a few hours, thanks for spotting this

**Comment by sdroege** [comment]
I fixed that now, but also see the discussion at https://github.com/rust-lang/rust/pull/47126#issuecomment-355149381 and the following 3 comments. It might make sense to reconsider the whole `TrustedRandomAccess` iterator or at least the zip implementation based on it.

**Comment by sdroege** [comment]
I think we're good to go actually

**Comment by sdroege** [comment]
Some benchmark results, based on the code from https://github.com/rust-lang/rust/issues/47115#issuecomment-354715511 . Just running a 1920\*1080\*4 byte array through the normal chunks version.

Something like 1.5% difference (the new version is faster), but consistently. So even for things that llvm can't auto-vectorize it seems to be slightly faster.

Before
```
$ cargo +local bench
running 1 test
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/deps/testing-611256b07a3291e9

test tests::test_bgrx_to_grey ... bench:   5,858,100 ns/iter (+/- 160,940)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out

$ cargo +local bench
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/deps/testing-611256b07a3291e9

running 1 test
test tests::test_bgrx_to_grey ... bench:   5,842,103 ns/iter (+/- 143,723)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out

$ cargo +local bench
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/deps/testing-611256b07a3291e9

running 1 test
test tests::test_bgrx_to_grey ... bench:   5,858,885 ns/iter (+/- 193,095)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out

$ cargo +local bench
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/deps/testing-611256b07a3291e9

running 1 test
test tests::test_bgrx_to_grey ... bench:   5,853,836 ns/iter (+/- 184,774)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out
```

After
```
$ cargo +local bench
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/deps/testing-611256b07a3291e9

running 1 test
test tests::test_bgrx_to_grey ... bench:   5,764,354 ns/iter (+/- 237,465)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out

$ cargo +local bench
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/deps/testing-611256b07a3291e9

running 1 test
test tests::test_bgrx_to_grey ... bench:   5,764,800 ns/iter (+/- 298,007)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out

$ cargo +local bench
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/deps/testing-611256b07a3291e9

running 1 test
test tests::test_bgrx_to_grey ... bench:   5,780,834 ns/iter (+/- 234,143)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out

$ cargo +local bench
    Finished release [optimized] target(s) in 0.0 secs
     Running target/release/deps/testing-611256b07a3291e9

running 1 test
test tests::test_bgrx_to_grey ... bench:   5,783,144 ns/iter (+/- 242,745)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out
```

**Comment by kennytm** [comment]
Thanks again! Just wanna confirm, will the expression `i * self.chunk_size` overflow?

**Comment by sdroege** [comment]
> Thanks again! Just wanna confirm, will the expression i * self.chunk_size overflow?

Yes, but only if you call it with an `i` that is not in range for the slice. AFAIU this is fine, that's what the "unchecked" / unsafe part of the trait is about: you need to ensure beforehand that the item you want to access does actually exist.

**Comment by sdroege** [comment]
Thanks a lot @kennytm and @bluss :)

**Pull request merged**